packages/transactions/src/decompile-transaction.ts

```diff
@@ -0,0 +1,181 @@
+
+function getAccountMetas(message: CompiledMessage): IAccountMeta[] {
+    const { header } = message;
+    const numWritableSignerAccounts = header.numSignerAccounts - header.numReadonlySignerAccounts;
+    const numWritableNonSignerAccounts =
+        message.staticAccounts.length - header.numSignerAccounts - header.numReadonlyNonSignerAccounts;
+
+    const accountMetas: IAccountMeta[] = [];
+
+    let accountIndex = 0;
+    for (let i = 0; i < numWritableSignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.WRITABLE_SIGNER,
+        });
+        accountIndex++;
+    }
+
+    for (let i = 0; i < header.numReadonlySignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.READONLY_SIGNER,
+        });
+        accountIndex++;
+    }
+
+    for (let i = 0; i < numWritableNonSignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.WRITABLE,
+        });
+        accountIndex++;
+    }
+
+    for (let i = 0; i < header.numReadonlyNonSignerAccounts; i++) {
+        accountMetas.push({
+            address: message.staticAccounts[accountIndex],
+            role: AccountRole.READONLY,
+        });
+        accountIndex++;
+    }
+
+    return accountMetas;
+}
+
+function convertInstruction(
+    instruction: CompiledMessage['instructions'][0],
+    accountMetas: IAccountMeta[]
+): IInstruction {
+    const programAddress = accountMetas[instruction.programAddressIndex]?.address;
+    if (!programAddress) {
+        // TODO coded error
+        throw new Error(`Could not find program address at index ${instruction.programAddressIndex}`);
+    }
+
+    const accounts = instruction.accountIndices?.map(accountIndex => accountMetas[accountIndex]);
+    const { data } = instruction;
+
+    return {
+        programAddress,
+        ...(accounts && accounts.length ? { accounts } : {}),
+        ...(data && data.length ? { data } : {}),
+    };
+}
+
+type LifetimeConstraint =
+    | {
+          blockhash: Blockhash;
+          lastValidBlockHeight: bigint;
+      }
+    | {
+          nonce: Nonce;
+          nonceAccountAddress: Base58EncodedAddress;
+          nonceAuthorityAddress: Base58EncodedAddress;
+      };
+
+function getLifetimeConstraint(
+    messageLifetimeToken: string,
+    firstInstruction?: IInstruction,
+    lastValidBlockHeight?: bigint
+): LifetimeConstraint {
+    if (!firstInstruction || !isAdvanceNonceAccountInstruction(firstInstruction)) {
+        // first instruction is not advance durable nonce, so use blockhash lifetime constraint
+        return {
+            blockhash: messageLifetimeToken as Blockhash,
+            lastValidBlockHeight: lastValidBlockHeight ?? 2n ** 64n - 1n, // U64 MAX
+        };
```

Comment:
I don't think that's possible because AFAICT a nonce account value is a valid blockhash. The value already looks suspicious: `7Rm8cnNwczto8KveNUjAEePjPHrtBvkSGPCJfzFJrLNB`

But then if you look at a nonce account on Explorer, it rather suspiciously labels it as "Blockhash": https://explorer.solana.com/address/8Two3Fmi24JA4woBUpRxEVjpeDsRPNJm3bLzPMiSChtk?cluster=devnet

So I don't think it's possible to distinguish them on the value of the lifetime token unfortunately!

Reply:
Ahhh yeah I remember you mentioning this in the transaction compat stuff. Smh.